linux_os/guide/system/permissions/files/permissions_within_important_dirs/root_permissions_syslibrary_files/ansible/shared.yml

```diff
@@ -0,0 +1,10 @@
+# platform = multi_platform_sle
+# reboot = false
+# strategy = restrict
+# complexity = high
+# disruption = medium
+
+- name: "Set ownership to root of system-wide library files"
+  command: "find {{ item }} ! -group root -type f -exec chgrp root '{}' \\;"
```

Comment:
afaik we are trying to avoid usage of command and shell if possible. Can this be possibly rewritten to use the file module?

Reply:
@vojtapolasek - Thank you for your proposal, but before I do some changes here I would like to clarify something. In current code the find command finds/collects the files without a correct group ownership and changes it to the correct one We need 1 loop. The second loop is nested indirectly. If we want to rewrite this code in ansible and to avoid bash - we will have 2 loops. With the 1st loop we will collect the files and with the 2nd we will apply the change on them.  If you can propose me a solution/sample with 1 loop with ansible I will adapt the current code to it. I think that in real situations we have to implement the rules asap - imagine a data center with 10K+ servers and more i.e. the speed of rules' implementation will play important role. The realization in ansible will be slower than the realization in bash. If for bash we need 5 sec and for ansible 2 seconds more - in case with 10K servers we will need only for this rule  333 minutes - 5.55h more!Are we looking for speed and efficiency or for implementation which will be based on a specific language? Any feedback will be welcomed - because I used this approach in a few rules and in this case it will be necessary to change my realization(this part of rule's code) in all of them.

Reply:
I think this might work:
https://docs.ansible.com/ansible/latest/collections/ansible/builtin/file_module.html
```
- name: Recursively change ownership of a directory
  ansible.builtin.file:
    path: {{ item }}
    state: directory
    recurse: yes
    group: root
```
I understand your concerns. Can you test this?

Reply:
@vojtapolasek  - According to the rule - we have to change the group ownership of a *file*  not of a directory => we have to collect all files in the /lib /lib64 /usr/lib and /usr/lib64 with owner != root and for them to apply change the group owner to root. The ansible file - doesn't provide ability to collect files. The ansible find - provides ability to collect files in a dictionary, but this dictionary does not have an attribute group. If you  can advice me how from a dictionary created by find to extact a file and to change its group - it will solve the problem. I was trying with this - and the result   files_matched - has a very complex structure 
 
- name: "Find system-wide library files"
  find:
    paths: "{{ item }}"
    file_type: "file"
    hidden: yes
  register: files_matched  
  with_items: [ '/lib' , '/lib64', '/usr/lib', '/usr/lib64' ]

https://docs.ansible.com/ansible/latest/collections/ansible/builtin/find_module.html

Reply:
Hi, I am sorry, I misunderstood you before. You are right, the file module will set ownership for directories and files. I think using the command module is therefore fine in this case.